core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -351,28 +321,25 @@ protected final void onSendingBytes(int numBytes) {
   protected final void onSentBytes(int numBytes) {
     boolean doNotify;
     synchronized (onReadyLock) {
+      boolean belowThresholdBefore = numSentBytesQueued < onReadyThreshold;
       numSentBytesQueued -= numBytes;
-      doNotify = needToNotifyOnReady();
+      boolean belowThresholdAfter = numSentBytesQueued < onReadyThreshold;
+      doNotify = !belowThresholdBefore && belowThresholdAfter;
     }
     if (doNotify) {
-      listener().onReady();
+      notifyIfReady();
     }
   }
 
-  /**
-   * Determines whether or not we need to call the {@link StreamListener#onReady()} handler now.
-   * Calling this method has the side-effect of unsetting {@link #shouldNotifyOnReady} so the
-   * handler should always be invoked immediately after calling this method.
-   */
-  @GuardedBy("onReadyLock")
-  private boolean needToNotifyOnReady() {
-    if (shouldNotifyOnReady && isReady()) {
-      // Returning true here counts as a call to the onReady callback, so
-      // unset the flag.
-      shouldNotifyOnReady = false;
-      return true;
+  @VisibleForTesting
+  final void notifyIfReady() {
```

Comment:
Are we sure we want `notifyIfReady()`. Calling `isReady()` instead of calculating `belowThresholdAfter` in `onSentBytes()` seems fine. And call to `isReady()` in `onStreamAllocated()` seems fine. And then in both cases outside of the synchronized block we call the listener.

Is there a deficiency to that?